### PR TITLE
Switch to semantic logger

### PR DESCRIPTION
**Pull request by segiddins**
It is more flexible than lograge, since we can use it for JSON logging everywhere, not just for requests

We have also mirrored the existing lograge config so this should be a strict superset

Development after:

image

Production after

```json
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:13.403606Z","level":"trace","level_index":0,"pid":90279,"thread":"SemanticLogger::Appenders","name":"SemanticLogger::Appenders","message":"Async: Appender thread active"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:13.427669Z","level":"info","level_index":2,"pid":90279,"thread":"7980","name":"Rails","message":"** [Honeybadger] Initializing Honeybadger Error Tracker for Ruby. Ship it! version=5.2.0 framework=rails level=1 pid=90279 -- honeybadger"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:13.427741Z","level":"warn","level_index":3,"pid":90279,"thread":"7980","name":"Rails","message":"** [Honeybadger] Development mode is enabled. Data will not be reported until you deploy your app. level=2 pid=90279 -- honeybadger"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:13.958853Z","level":"trace","level_index":0,"pid":90293,"thread":"7980","name":"SemanticLogger::Appenders","message":"Reopening appender: SemanticLogger::Appender::File"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:13.959555Z","level":"trace","level_index":0,"pid":90293,"thread":"7980","name":"SemanticLogger::Appenders","message":"All appenders re-opened"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:13.959815Z","level":"trace","level_index":0,"pid":90293,"thread":"SemanticLogger::Appenders","name":"SemanticLogger::Appenders","message":"Async: Appender thread active"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:25.930506Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"Rails","message":"Started HEAD \"/fdsfs\" for 127.0.0.1 at 2023-03-02 18:23:25 -0800"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:25.939099Z","level":"fatal","level_index":5,"pid":90293,"thread":"puma srv tp 001","file":"/Users/segiddins/.gem/ruby/3.2.1/gems/concurrent-ruby-1.2.2/lib/concurrent-ruby/concurrent/atomic/thread_local_var.rb","line":93,"name":"Rails","exception":{"name":"ActionController::RoutingError","message":"No route matches [HEAD] \"/fdsfs\"","stack_trace":["/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/debug_exceptions.rb:32:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/show_exceptions.rb:26:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/railties-7.0.4.2/lib/rails/rack/logger.rb:40:in `call_app'","/Users/segiddins/.gem/ruby/3.2.1/gems/railties-7.0.4.2/lib/rails/rack/logger.rb:25:in `block in call'","/Users/segiddins/.gem/ruby/3.2.1/gems/semantic_logger-4.12.0/lib/semantic_logger/base.rb:190:in `block in tagged'","/Users/segiddins/.gem/ruby/3.2.1/gems/semantic_logger-4.12.0/lib/semantic_logger/semantic_logger.rb:347:in `tagged'","/Users/segiddins/.gem/ruby/3.2.1/gems/semantic_logger-4.12.0/lib/semantic_logger/base.rb:202:in `tagged'","/Users/segiddins/.gem/ruby/3.2.1/gems/railties-7.0.4.2/lib/rails/rack/logger.rb:25:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/sprockets-rails-3.4.2/lib/sprockets/rails/quiet_assets.rb:13:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/remote_ip.rb:93:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/request_id.rb:26:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/rack-2.2.6.2/lib/rack/method_override.rb:24:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/rack-2.2.6.2/lib/rack/runtime.rb:22:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/activesupport-7.0.4.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/server_timing.rb:61:in `block in call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/server_timing.rb:26:in `collect_events'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/server_timing.rb:60:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/executor.rb:14:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/static.rb:23:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/rack-2.2.6.2/lib/rack/sendfile.rb:110:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/actionpack-7.0.4.2/lib/action_dispatch/middleware/host_authorization.rb:137:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/rack-utf8_sanitizer-1.8.0/lib/rack/utf8_sanitizer.rb:28:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/ddtrace-1.9.0/lib/datadog/tracing/contrib/rack/middlewares.rb:88:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/honeybadger-5.2.0/lib/honeybadger/rack/error_notifier.rb:33:in `block in call'","/Users/segiddins/.gem/ruby/3.2.1/gems/honeybadger-5.2.0/lib/honeybadger/agent.rb:433:in `with_rack_env'","/Users/segiddins/.gem/ruby/3.2.1/gems/honeybadger-5.2.0/lib/honeybadger/rack/error_notifier.rb:30:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/honeybadger-5.2.0/lib/honeybadger/rack/user_feedback.rb:31:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/honeybadger-5.2.0/lib/honeybadger/rack/user_informer.rb:21:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/railties-7.0.4.2/lib/rails/engine.rb:530:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/railties-7.0.4.2/lib/rails/railtie.rb:226:in `public_send'","/Users/segiddins/.gem/ruby/3.2.1/gems/railties-7.0.4.2/lib/rails/railtie.rb:226:in `method_missing'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/configuration.rb:269:in `call'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/request.rb:98:in `block in handle_request'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/thread_pool.rb:340:in `with_force_shutdown'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/request.rb:97:in `handle_request'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/server.rb:431:in `process_client'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/server.rb:233:in `block in run'","/Users/segiddins/.gem/ruby/3.2.1/gems/puma-6.1.1/lib/puma/thread_pool.rb:147:in `block in spawn_thread'"]}}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:25.944678Z","level":"error","level_index":4,"pid":90293,"thread":"puma srv tp 001","file":"/Users/segiddins/.gem/ruby/3.2.1/gems/honeybadger-5.2.0/lib/honeybadger/logging.rb","line":98,"name":"Honeybadger","message":"Unable to send error report: API key is missing. id=0abfe61f-5df7-4bfc-83b8-bb8be93ba14c level=3 pid=90293 -- honeybadger"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.128361Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"Rails","message":"Started GET \"/\" for 127.0.0.1 at 2023-03-02 18:23:29 -0800"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.133418Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"Rails","message":"Processing by HomeController#index as */*"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.186145Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"ActionView::Base","message":"Rendered home/index.html.erb within layouts/application (Duration: 2.4ms | Allocations: 1738)"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195087Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"ActionView::Base","message":"Rendered layout layouts/application.html.erb (Duration: 11.4ms | Allocations: 10280)"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195380Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"Rails","message":"Completed 200 OK in 62ms (Views: 13.4ms | ActiveRecord: 13.4ms | Allocations: 17305)"}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195606Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"StatsD","message":"measure rails.performance.total_duration","payload":{"type":"measure","name":"rails.performance.total_duration","value":61.91799998283386,"sample_rate":1.0,"tags":{"controller":"HomeController","action":"index","format":"all","env":"development"}}}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195656Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"StatsD","message":"measure rails.performance.db_time","payload":{"type":"measure","name":"rails.performance.db_time","value":13.439999908208847,"sample_rate":1.0,"tags":{"controller":"HomeController","action":"index","format":"all","env":"development"}}}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195678Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"StatsD","message":"measure rails.performance.view_time","payload":{"type":"measure","name":"rails.performance.view_time","value":13.371999986702576,"sample_rate":1.0,"tags":{"controller":"HomeController","action":"index","format":"all","env":"development"}}}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195704Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"StatsD","message":"histogram rails.performance.allocations","payload":{"type":"histogram","name":"rails.performance.allocations","value":17363,"sample_rate":1.0,"tags":{"controller":"HomeController","action":"index","format":"all","env":"development"}}}
{"host":"X-Wing.haashold","application":"Semantic Logger","environment":"development","timestamp":"2023-03-03T02:23:29.195720Z","level":"info","level_index":2,"pid":90293,"thread":"puma srv tp 001","name":"StatsD","message":"increment rails.performance.status.200","payload":{"type":"increment","name":"rails.performance.status.200","value":1,"sample_rate":1.0,"tags":{"controller":"HomeController","action":"index","format":"all","env":"development"}}}
```

Development before:
image

Production before:
image

### CODE REVIEW

**Comment by indirect** [comment]
Nice! Can you show us some before and after output samples from local development and production?

**Comment by simi** [comment]
How painful it would be to keep common development `bin/rails s` output same as vanilla Rails app?

**Comment by segiddins** [comment]
It would be easy, but I personally strongly prefer the highlighting from semantic logger

**Comment by simi** [comment]
> It would be easy, but I personally strongly prefer the highlighting from semantic logger

Maybe you can enable it somehow easily locally? I usually prefer plain Rails output (TBH I don't really care actually), but it is much more friendly to new contributors as well IMHO.

**Comment by indirect** [comment]
For development mode logging, can we turn on semantic logging using an env var? I also think we should keep the default Rails logs by default in development because that is what everyone already knows how to read.

I'm also a little worried about switching to semantic logger given that it blocks upgrading to Rails 7.1: https://github.com/reidmorrison/semantic_logger/issues/264.

**Comment by segiddins** [comment]
Give this another whirl now, development by default should look just like it did before this change!

File: config/environments/development.rb [hunk 1]
```diff
@@ -87,8 +100,12 @@
     config.cache_classes = true
     config.eager_load = true
 
-    config.logger = ActiveSupport::Logger.new($stdout)
     config.log_level = :info
+    config.rails_semantic_logger.format     = :json
+    config.rails_semantic_logger.semantic   = true
+    config.rails_semantic_logger.started    = false
+    config.rails_semantic_logger.processing = false
+    config.rails_semantic_logger.rendered   = false
```
Comment (indirect) [inline config/environments/development.rb:108]: am I misreading this, or does this override the section above and force it back to JSON regardless of the ENV var?

File: config/initializers/delayed_job.rb [hunk 2]
```diff
@@ -2,6 +2,5 @@
 Delayed::Worker.max_attempts = 10
 Delayed::Worker.max_run_time = 5.minutes
 Delayed::Worker.destroy_failed_jobs = false
-Delayed::Worker.logger = Logger.new(Rails.root.join('log', 'delayed_job.log'))
```
Comment (indirect) [inline config/initializers/delayed_job.rb:5]: does DJ also need a `SemanticLogger[DelayedJob]` here?

File: app/models/gem_download.rb [hunk 3]
```diff
@@ -1,4 +1,6 @@
 class GemDownload < ApplicationRecord
+  include SemanticLogger::Loggable
```
Comment (indirect) [inline app/models/gem_download.rb:2]: can we make ApplicationRecord's default logger be the semantic one, so we don't have to one-off include the module in individual record classes?

File: app/jobs/indexer.rb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 class Indexer < ApplicationJob
   extend StatsD::Instrument
   include TraceTagger
+  include SemanticLogger::Loggable
```
Comment (indirect) [inline app/jobs/indexer.rb:4]: Ditto for the ApplicationJob logger so we don't have to include the semantic module in each individual job class, I think?

**Review by indirect — APPROVED** [review]
Looks good! A couple questions about config but I'm happy to have parsable logs in production 👍🏻

File: config/initializers/delayed_job.rb (hunk 2 above)
Comment (segiddins) [reply config/initializers/delayed_job.rb:5]: semantic_logger_rails has automatic support for DJ

File: config/environments/development.rb (hunk 1 above)
Comment (segiddins) [reply config/environments/development.rb:108]: only if `ENV["PROFILE"]` is set

**Pull request merged**